src/shogun/mathematics/linalg/BaseVector.h

```diff
@@ -54,7 +54,7 @@ struct BaseVector
 	}
 
 	/** Data Storage */
-	virtual bool onGPU() = 0 ;
+	virtual bool onGPU() const = 0 ;
```

Comment:
well, this was decided because dynamic cast is thought to be slower compared to one virtual call + static cast. But I see what you mean. We probably won't need to have two different classes, only one class should suffice.